### ISSUE TITLE
Can 3DGS render depth maps?

### ISSUE THREAD

**Issue by Lizhinwafu** [graphdeco-inria/gaussian-splatting#983]
Can 3DGS render depth maps?

**Comment by jaco001** [comment]
This is 'almost' point cloud so you can render depth map. (eg https://www.techrxiv.org/users/687263/articles/680135-three-to-two-a-novel-approach-for-converting-3d-point-clouds-into-2d-depth-maps )

**Comment by MultiTrickFox** [comment]
Yes it can do depth calculations in this repo's own render function

**Comment by MultiTrickFox** [comment]
Yes the renderer takes in a view direction, outputs the produced image as well as depth from that view.
It can be any view, in case of trainer its a training view, but it can be anything.
look at here: ```https://github.com/graphdeco-inria/gaussian-splatting/blob/c643c864416c5258c76cd9f0eed55203834cbf83/train.py```
line 94 and 106.

**Comment by jlartois** [comment]
For anyone else that wants to render depth maps during training, you need to switch to the dev branch. You will indeed get access to `invDepth = render_pkg["depth"]`, for example in train.py. This comes from `submodules/diff-gaussian-rasterization/diff_gaussian_rasterization/__init__.py` (note that this is the `dr_aa` branch of that submodule). 

However, how do we go from `invDepth` to the actual depth? Just invert it?

**Comment by jlartois** [comment]
I tried some things to get rid of the splats that were negatively impacting the depth map. I pruned away splats based on a combination of:
- the scaling: prune large splats
- opacity: prune splats with a low opacity
- position: prune splats close to the camera for which you are rendering the depth map

The resulting depth map (shown below) is much more accurate. I.e., when I use it to warp the view to a neighboring view, the image loss is very low.
![truck_000108](https://github.com/user-attachments/assets/af00e8e3-a5f1-4c3d-ac6b-3c8fcb5151af)

Side note: I think this means that the current implementation of depth normalization is flawed (unless you can reliably prune the incorrect splats).

**Comment by leblond14u** [comment]
Hi, 

Have you tried disabling the alpha blending process for the depth rendering ? 
Maybe this could yield better results ? 

Best, 

Hugo

**Comment by Lizhinwafu** [comment]
NeRF model can render better depth？





------------------ Original ------------------
From: HugoLbld ***@***.***&gt;
Date: Tue,Nov 12,2024 11:42 AM
To: graphdeco-inria/gaussian-splatting ***@***.***&gt;
Cc: Lizhi Jiang ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [graphdeco-inria/gaussian-splatting] Can 3DGS render depth maps?(Issue #983)





 
Hi,
 
Have you tried disabling the alpha blending process for the depth rendering ?
 Maybe this could yield better results ?
 
Best,
 
Hugo

**Comment by jlartois** [comment]
@leblond14u, If you disable the alpha blending, how do you determine the depth of each pixel?

@Lizhinwafu, depth maps rendered from NeRF models generally are good. As with most light field techniques, the depth in textureless areas can be noisy though.

**Comment by leblond14u** [comment]
@jlartois You ca still determine the depth value for every pixels by measuring the depth to the first touched Gaussian. 
In theory to get the best result possible the gaussians are supposed to arrange themselves on the represented surface and once aligned get an opacity close to 1. So following this idea training a GS without alpha blending on the depth part could potentially improve the depth results. Papers like **SuGaR** are explaining this theory in a clearer way if you need some references.

**Comment by Egolee0713** [comment]
> @jlartois You ca still determine the depth value for every pixels by measuring the depth to the first touched Gaussian. In theory to get the best result possible the gaussians are supposed to arrange themselves on the represented surface and once aligned get an opacity close to 1. So following this idea training a GS without alpha blending on the depth part could potentially improve the depth results. Papers like **SuGaR** are explaining this theory in a clearer way if you need some references.

@leblond14u It looks like a good idea. How can I do this in this repo or where is the corresponding gaussian-rasterization impletemention.

**Comment by ali-john** [comment]
@Egolee0713 Were you able to extract depths? I think the depth maps are poor because they are alpha blended. If we can mark the first interaction with gaussian as depth, the depth maps might improve. I think we have to change the forward.cu kernel for this. 
@leblond14u is this the correct approach?

Thanks.
